<html>
<head>
<title>Mandatory Error Codes Revisited</title>
<link rel="Stylesheet" rev="Stylesheet" href="../../../../forms/Layout.css" type="text/css">
<link rel="Stylesheet" rev="Stylesheet" href="../../../../forms/FontStyles.css" type="text/css">
<link rel="Stylesheet" rev="Stylesheet" href="../../../../forms/newarticle.css" type="text/css">
<script src="../../../../forms/popwindow.js"></script>
</head>

<body BGCOLOR="#ffffff" LINK="#0000ff" VLINK="#330066" ALINK="#ff0000" TEXT="#000000">
<!--Copyright &#169; Dr. Dobb's Journal-->
<p><i>Dr. Dobb's Journal</i> September 2006</p>

<h1>Mandatory Error Codes Revisited</h1>
<h2>Forcing the use of error codes</h2>

<h3>by Guy Peleg</h3>

<I>Guy is chief architect of event  processing for the Amdocs Billing Platform (<a href="http://www.amdocs.com">http://www.amdocs.com</a>). He can be contacted at guy.peleg@amdocs.com.</I>


<hr>




<p>In his article "Three Ideas" (<a href="http://www.ddj.com/dept/cpp/184401917">http://www.ddj.com/dept/cpp/184401917</a>), Andrei Alexandrescu described a framework for "mandatory error codes" that forces callers of a method to accept and use the returned error code. In this article, I refer to these codes as "forcing methods." </p>

<p>The idea is simple, yet powerful: A template class <i>ErrorCode</i>, which holds the real error code, is returned from a forcing method. When the returned <i>ErrorCode</i> goes out of scope, it throws an exception&#151;unless the caller uses the error code (using the cast operator). The caller can explicitly ignore the error code by casting the <i>ErrorCode</i> with <i>IgnoreError</i>. Listing One presents the classes that make up this framework, while Example 1 shows how you can use it.</p>


<pre class="code">


struct IgnoreError {};

template &lt;class T&gt;
class ErrorCode
{
    mutable bool    read_;
    T           code_;
public:
    ErrorCode(const T&amp; code):
read_(false), code_(code) {
}
ErrorCode(const ErrorCode&amp; rhs):
read_(rhs.read_), code_(rhs.code_) {
    rhs.read_ = true;
}
operator T() {
    // disarm exception
    this-&gt;read_ = true;
    return this-&gt;code_;
}
    operator IgnoreError()  {
    // disarm exception
    this-&gt;read_ = true;
    return IgnoreError();
}
~ErrorCode() {
    if (!this-&gt;read_)
    {
        throw MandatoryErrorCodeException(....);
    }
}
};</pre>
<div class="caption">
   Listing One
</div>	
	
	


<div>

<pre class="code">
ErrorCode&lt;int&gt; FallableFunction();	// function declaration
ErrorCode&lt;int&gt; result = FallableFunction();	// ok
if (FallableFunction ()) {...}	// ok
FallableFunction();	// will throw
(IgnoreError) FallableFunction();	// ok
</pre>

<div class="caption">
Example 1: Using the framework.
</div>
</div>




<p>However, the problem with this framework is the exception that might be thrown at the destructor of <i>ErrorCode</i>. For instance, Example 2 looks perfectly normal. You don't ignore the returned error codes; you accept and use them. There's even a special <i>recover </i>method that can handle everything. Well, almost everything.</p>


<div>

<pre class="code">
void AClass::someMethod() 
{
   try {
      ErrorCode&lt;int&gt; ec1 = this-&gt;fallableMethod_1();
      ErrorCode&lt;int&gt; ec2 = this-&gt;fallableMethod_2();
      if (ec1 &amp;&amp; ec2) {
	  this-&gt;doSomething();
      }
   }
   catch (...) {
      this-&gt;recover();
   }
}
</pre>

<div class="caption">
Example 2: Problem code.
</div>
</div>


<p>Here's the problem: If <i>this-&gt;fallableMethod_2()</i> throws, then the <i>ec1</i> destructor is activated. But since no one has yet used <i>ec1</i>, another exception is thrown (from activating <i>ec1</i>'s destructor), which terminates the process (<i>set_terminate</i> is called but this won't help). This is a big surprise if you've followed all the rules.</p>






<h3>The Solution</h3>

<p>If you want to use this kind of framework effectively, you first need to tweak it a bit.</p>

<p>First, change your assumption that an <i>ErrorCode</i> must be used. There is no real way to make sure that users actually used it. You should be more concerned with callers accepting the return code, rather than having them use it. After all, a caller to this kind of method can accept the error code, then assign it to some variable that is not used.</p>

<p>Second, make sure that implicitly thrown exceptions won't terminate the process. To do this, first add a new class, <i>ThrowableErrorCode</i> (Listing Two). This class is returned from methods that want to enforce error code checking, instead of the <i>ErrorCode</i>. <i>ThrowableErrorCode</i> doesn't grant access to the error code it carries.</p>

	

<pre class="code">
template &lt;class CODE&gt;
class ThrowableErrorCode

{
    template &lt;class CODE&gt; friend class ErrorCode;
    CODE        code_;
    bool        throw_;
public:
    /// ctor - receive the code and arm the exception
    ThrowableErrorCode(CODE i_code):
    code_(i_code),
    throw_(true)
    {}
    // explicitly ignore the error code and avoid exception
    operator IgnoreError() {
        this-&gt; throw_ = false;
        return IgnoreError();
    }
    ~ThrowableErrorCode() {
        // will throw unless ErrorCode&lt;CODE&gt; or
                // IgnoreError will prevent it
        if (this-&gt; throw_)
        {
            throw MandatoryErrorCodeException(...);
        }
    }
};</pre>
<div class="caption">
   Listing Two
</div>	
	
	
	


<p>As its name implies, <i>ThrowableErrorCode</i> might throw an exception. In fact, it throws one when it goes out of scope. The only way to stop it from doing so is by either explicitly ignoring the error code using <i>ThrowableErrorCode</i>'s operator <i>IgnoreError</i>, or by assigning it into an <i>ErrorCode</i>.</p>

<p>The new <i>ErrorCode</i> (Listing Three) does not throw an exception. It disarms <i>ThrowableErrorCode</i> and grants access to the error code. </p>

	

<pre class="code">
template &lt;class CODE&gt;
class ErrorCode {
    CODE    code_;
public:
    // Explicit ctor to make sure that the user of this
    // class knows what she/he is doing
    explicit ErrorCode(ThrowableErrorCode&lt;CODE&gt;&amp; code):
    code_(code.code_) {
        // prevent the throw
        code.throw_ = false;
    }
    operator CODE() {
        return this-&gt;code_;
    }
    ~ErrorCode()
    {}
};</pre>
<div class="caption">
   Listing Three
</div>	
	

<p>So the caller to this kind of method has these alternatives:</p>



<ul>
  <li>Accept the returned <i>ThrowableErrorCode</i> into an <i>ErrorCode</i> (which, presumably, the caller uses). Good practice. </li>
  <li>Explicitly ignore the returned error code by using the <i>IgnoreError</i> operator. Good practice. </li>
  <li>Ignore the returned <i>ThrowableErrorCode</i>, in which case a temporary instance of <i>ThrowableErrorCode</i> is created and deleted within the scope of a statement, which causes an exception but won't jeopardize the process. Bad practice. </li>
  <li>Accept the return code as <i>ThrowableErrorCode</i>, which means that the exception is not disarmed. This means that callers are willingly holding an instance of a class that might throw, in which case this caller better know what they are doing. Bad practice. </li>
</ul>



<p>Example 3 shows how you can use this new framework. Now when you revisit the problem encountered in Example 2, the code works fine because <i>ErrorCodes</i> <i>ec1</i> and <i>ec2</i> disarm the <i>ThrowableErrorCodes</i> that return from the fallable methods. That is, in case of an exception from these methods, the code reaches the <i>catch</i> statement just as you intended.</p>


<div>

<pre class="code">
// function declaration
ThrowableErrorCode&lt;int&gt; FallableFunction(); 	


// ok
ErrorCode&lt;int&gt; result = FallableFunction();
if (0 == result) {...}	

// will throw - but will not terminate the process
if (FallableFunction ()) {...} 

// will throw - but will not terminate the process
FallableFunction(); 

// ok - Explicitly ignoring the error code
(IgnoreError) FallableFunction(); 

// might throw - the caller is taking a risk!
ThrowableErrorCode&lt;int&gt; mightThrow = FallableFunction(); 
</pre>

<div class="caption">
Example 3: Example use of the new framework.
</div>
</div>





<h3>Acknowledgment</h3>

<p>Thanks to Andrei Alexandrescu.</p>
 <p>
DDJ
</p>

</body>
</html>